### PR TITLE
Remove class constraint from Interlocked.{Compare}Exchange

### PR DESCRIPTION
**Pull request by stephentoub**
https://github.com/dotnet/runtime/issues/65184

### CODE REVIEW

**Comment by stephentoub** [comment]
@MihuBot

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs
Comment (MichalPetryka) [inline src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:110]: This will change the behaviour for bools and floating point I think.

File: src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs
Comment (MichalPetryka) [inline src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: I assume bool overloads being slower on some platforms doesn't matter here?

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (MichalPetryka) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: This should probably get re-added to the JIT intrinsic.

File: src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs
Comment (stephentoub) [reply src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs:110]: T is constrained to `IBinaryInteger<T>`.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: I likely will, but wanted to see the impact on codegen

File: src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs
Comment (stephentoub) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: Which platforms?

Comment (MichalPetryka) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: RISC-V will be significantly impacted since their Atomics ISA only provides 4B and 8B atomics so 1B and 2B have to be emulated (they've recently ratified a separate ISA for 1B and 2B atomics but it'll take a while for hardware to get it). Other platforms all have small atomics I think but only X86/X64 and ARM64 (and ARM32 in #97792) have it implemented today and even on those platforms they're slightly more expensive due to requiring zero-extensions.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: If you want to optimize it for CoreCLR, putting the non `class` path in a separate local method might be slightly better for avoiding regressions with the inliner after you re-add `Intrinsic` since it should then almost never be reached (only for debug and indirect calls I think).

File: src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs
Comment (stephentoub) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: Thanks. That seems fine. As noted in the PR where the new Interlocked APIs were added, if we're happy to expose APIs for the smaller data types, we should be happy to use them everywhere they're applicable. Otherwise we're just creating a pit of failure by exposing APIs that are more attractive / relevant but then saying they shouldn't be used.

Comment (EgorBo) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: Perhaps, by the time RISC-V gets more popular it will get an extension to do byte atomics natively, for now, it feels a bit odd to trade readability over peformance (overhead is not clear, very much could be invisible) on a single non-mainstream platforms to be fair. 

Example: https://godbolt.org/z/dovM1va3v

Comment (stephentoub) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: > it feels a bit odd to trade readability over peformance

Meaning you agree using bool instead of int is desirable, yes?

Comment (EgorBo) [reply src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/Marshalling/ComObject.cs:26]: Well, we either never use `Interlocked.X` for `bool` for the same reason (why did we expose it then?) or use everywhere where appropriate, but it's just my opinion, I don't know how bad the impact potentially is 🙂

File: src/coreclr/vm/jitinterface.cpp
Comment (MichalPetryka) [inline src/coreclr/vm/jitinterface.cpp:7265]: I don't think this will be needed if we re-apply the JIT intrinsic, only the object part might slightly help the inliner.

Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:7265]: Yep, we already discussed this, I think Stephen decided to leave it for separate clean up

Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7265]: @jkotas had also requested offline that either all cases be handled as intrinsic or nothing be intrinsic.

Comment (MichalPetryka) [reply src/coreclr/vm/jitinterface.cpp:7265]: > @jkotas had also requested offline that either all cases be handled as intrinsic or nothing be intrinsic.

JIT can't handle the object case as intrinsic for most cases due to the need for memory barriers.

Comment (jkotas) [reply src/coreclr/vm/jitinterface.cpp:7265]: > @jkotas had also requested offline that either all cases be handled as intrinsic or nothing be intrinsic.

My concern was that the generic method with type switches would be too much work for the inliner and it would produce sub-optimal code when used in real-world code without intrinsic handling. I assume that the codegen diffs captured earlier in this PR confirmed that. @stephentoub Is there a short summary of what you found in https://github.com/dotnet/runtime/pull/104558#discussion_r1669457451 ?

I do not have an opinion about where the intrinsic handling happens exactly.

> JIT can't handle the object case as intrinsic for most cases due to the need for memory barriers (it only handles null and FOH objects there).

Handle as intrinsic != produce code without any calls. There is nothing fundamental preventing all intrinsic IL rewriting done by the VM to be done in the JIT instead. (We may need some new methods on JIT/EE interface.)

File: src/coreclr/vm/jitinterface.cpp
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:7272]: This is static memory. You need to have arrays of these so that each primitive type gets its own copy.

File: src/coreclr/vm/jitinterface.cpp
Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:7265]: > My concern was that the generic method with type switches would be too much work for the inliner 

We can  workaround this by, basically, special-casing in JIT so it knows that this method consumes 0 out of JIT Inliner's budget. In theory, it wouldn't be needed if our inliner budget was smart enough to predict foldable branches (and their sizes) precisely.

Comment (jkotas) [reply src/coreclr/vm/jitinterface.cpp:7265]: > special-casing in JIT so it knows that this method consumes 0 out of JIT Inliner's budget

Would this special casing also need to include how this method gets imported so that it actually consumes 0 out of JIT inliners budget (e.g. no locals create by the inlining, etc.)?

Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:7265]: > > special-casing in JIT so it knows that this method consumes 0 out of JIT Inliner's budget
> 
> Would this special casing also need to include how this method gets imported so that it actually consumes 0 out of JIT inliners budget (e.g. no locals create by the inlining, etc.)?

No, but I'd expect most of its body to be trimmed early during importer (all those `if (Unsafe.Sizeof`) so those branches won't be even materialized. The problem that Inliner can't precisely predict it. There might be some extra locals compare to what we see now but I am not sure. I can try to handle this as a follow up if this is merged as is.

**Comment by timcassell** [comment]
Is there a reason not to support structs with size <= 8?

**Comment by EgorBo** [comment]
> Is there a reason not to support structs with size <= 8?

Our memory model doesn't guarantee atomicity for those (nor you can use them with Volatile.Read/Write). Same for e.g. field assignment in classes

**Comment by stephentoub** [comment]
> > Is there a reason not to support structs with size <= 8?
> 
> Our memory model doesn't guarantee atomicity for those (nor you can use them with Volatile.Read/Write). Same for e.g. field assignment in classes so while we might support them in this API, it won't be supported in other places

And equality may not be respected, eg a custom Equals wouldn't be used, and even if there isn't one, padding in the type could contain random bits that would make two otherwise equal values unequal.

File: src/coreclr/vm/jitinterface.cpp
Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7265]: > @stephentoub Is there a short summary of what you found in https://github.com/dotnet/runtime/pull/104558#discussion_r1669457451 ?

A bunch of cases where calls previously inlined were no longer inlined.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:256]: This can be `sizeof(T)`. It is cheaper than `Unsafe.SizeOf<T>`.

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs]: Should codegen lower one into the other (where possible)?

**Comment by stephentoub** [comment]
@jkotas, I appear to have messed up naot here. Suggestions on where to look? Did I miss somewhere else intrinsics need to be updated?

**Comment by MichalStrehovsky** [comment]
> @jkotas, I appear to have messed up naot here. Suggestions on where to look? Did I miss somewhere else intrinsics need to be updated?

It's hitting an infinite recursion because the implementation of `CompareExchange<T>` now depends on `CompareExchange<T>` (we need `CompareExchange<T>` to be able to do fundamental things like running static constructors, etc and this is doing reflection that is way higher up the stack.

I'm still looking why the IL wasn't replaced, but without replacing the IL, the new method body of `CompareExchange` is not executable. Do we expect the method body to actually execute? Shouldn't it be just a throw?

File: src/coreclr/vm/jitinterface.cpp
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:7260]:
```suggestion
            case 2: ilIndex = 2; break;
            case 4: ilIndex = 3; break;
            case 8: ilIndex = 4; break;
```

File: src/coreclr/vm/jitinterface.cpp
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:7266]: It would be better to move the statics into to switch statement so that we do not have to fetch 5 methods each time even though we really only need one.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (MichalPetryka) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:240]:
```suggestion
        [Intrinsic]
        [return: NotNullIfNotNull(nameof(location1))]
```
This should still get the JIT handling to reduce diffs.

**Review by MichalPetryka — CHANGES_REQUESTED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:240]:
```suggestion
        [Intrinsic]
        [return: NotNullIfNotNull(nameof(location1))]
```

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: I think this will require hardening expansion of the Exchange/CompareExchange in the JIT against unsupported `T` so that it only expands primitives and object refs and we fallback to throwing the NotSupportedException.

File: src/coreclr/vm/jitinterface.cpp
Comment (stephentoub) [reply src/coreclr/vm/jitinterface.cpp:7266]: Ok

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: The VM and the JIT only expands method marked as `[Intrinsic]`. I do not think any of the IL expansion is kicking in with the current version of the change.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: Yes, I'm not sure how I managed to delete that and I'll fix that. But with regards to hardening, is there something in particular you're seeing that looks wrong or you're just making a general comment? Regardless, I'll put back the attribute and fix the overload selection copy/paste error.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: I expect that you will hit this assert: https://github.com/dotnet/runtime/blob/db1f075c281b443a7de8b4523a2e1ea17dc39af6/src/coreclr/jit/importercalls.cpp#L4115 for the unsupported cases (we can wait for the CI to run to validate that it really the case). Changing that assert into `break;` similar to other cases should fix the problem.

Comment (MichalPetryka) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:491]: > I expect that you will hit this assert:
> 
> https://github.com/dotnet/runtime/blob/db1f075c281b443a7de8b4523a2e1ea17dc39af6/src/coreclr/jit/importercalls.cpp#L4115
> 
> for the unsupported cases (we can wait for the CI to run to validate that it really the case). Changing that assert into `break;` similar to other cases should fix the problem.

The `else if (!varTypeIsIntegral(retType))` should handle things there already.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:492]: > This should get a test that verifies the exception for structs though.

I'd already added one:
https://github.com/dotnet/runtime/pull/104558/files#diff-b7ba854effe0b40e624638bc9ea70e50950841a776d09a3a1f36df3922f9efdfR397

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (MichalPetryka) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:506]: Are we sure we want to accept floating point data here? Since the code does bitwise equality, having them here could be misleading.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:506]: > Are we sure we want to accept floating point data here? Since the code does bitwise equality, having them here could be misleading.

How is that different from the existing overloads strongly-typed to take float/double?
https://github.com/dotnet/runtime/blob/ac0eeb015a0266adb06696bb58c21e5bfe6d9b56/src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs#L366-L378

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs:506]: I would find it incredibly surprising if the behavior of `Interlocked.CompareExchange(ref float, ...)` and `Interlocked.CompareExchange<float>(ref float, ...)` differed. If we're talking about changing the behavior of both, sure, that's a breaking change that could be considered, but as you say it's not related to this PR.

File: src/coreclr/vm/corelib.h
Comment (jkotas) [inline src/coreclr/vm/corelib.h:700]:
```suggestion
DEFINE_METHOD(INTERLOCKED,          COMPARE_EXCHANGE_INT,  CompareExchange, SM_RefInt_Int_Int_RetInt)
DEFINE_METHOD(INTERLOCKED,          COMPARE_EXCHANGE_LONG, CompareExchange, SM_RefLong_Long_Long_RetLong)
```
It would make more sense to use the signed int and long since they are the actual FCalls. UInt and ULong is going to create an extra level of inlining for the JIT

**Comment by jkotas** [comment]
The JIT is failing with:
```
Assert failure(PID 55347 [0x0000d833], Thread: 55394 [0xd862]): Assertion failed '(genActualType(op2->TypeGet()) == genActualType(info.compRetType)) || ((op2->TypeGet() == TYP_I_IMPL) && (info.compRetType == TYP_BYREF)) || ((op2->TypeGet() == TYP_BYREF) && (info.compRetType == TYP_I_IMPL)) || (varTypeIsFloating(op2->gtType) && varTypeIsFloating(info.compRetType)) || (varTypeIsStruct(op2) && varTypeIsStruct(info.compRetType)) : Possibly bad IL with CEE_ret at offset 0008h (op1=NULL op2=long stkDepth=0)' in 'System.Threading.Interlocked:CompareExchange[double](byref,double,double):double' during 'Importation' (IL size 9; hash 0xde0a4bf1; Tier0)
```
It looks like a problem with the implicit double->long conversions.

File: src/coreclr/vm/jitinterface.cpp
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:7266]: Also reject CorInfoType::IsFloatingPoint here. (It would be nice to call `GetVerifierCorElementType` only once for both checks.)

File: src/coreclr/vm/corelib.h
Comment (jkotas) [reply src/coreclr/vm/corelib.h:700]: (similar change for nativeaot too)

File: src/coreclr/vm/corelib.h
Comment (jkotas) [inline src/coreclr/vm/corelib.h:698]: Nit: UShort/Short (and SByte/Byte) have signed wrapper and unsigned actual implementation, so it would be better to use UShort here.

Comment (stephentoub) [reply src/coreclr/vm/corelib.h:698]: I see that now and will revert the ushort changes... but why do we differ like between short and int/long?

Comment (jkotas) [reply src/coreclr/vm/corelib.h]: Historically, we used the signed actual implementation and unsigned wrapper everywhere.

It was flipped in https://github.com/dotnet/runtime/pull/99011/ for small integer types. The fallback implementation for unsigned types is a bit more natural to write (no need to worry about sign-extensions during bit twiddling).

**Comment by jkotas** [comment]
@lambdageek @steveisok Are there any Mono-specific concerns related to this change?

File: src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs]: The final code should be the same. `sizeof` is cheaper for startup. `sizeof` is IL opcode. ``Unsafe.SizeOf` is JIT intrinsic. It requires importing the call of the method, creating the generic method instantiation and recognizing the method by name as intrinsic, ... . None of this is necessary for the `sizeof` opcode.

**Comment by lambdageek** [comment]
> @lambdageek @steveisok Are there any Mono-specific concerns related to this change?

Are there new/existing tests for the throwing behavior?

If mono is green, one of two things is wrong (I'm on a phone and can't check):
1. Mono intrinsics are firing but they haven't been updated to throw if the arguments are non-primitive valuetypes, so they're working by accident
2. The intrinsics aren't firing and the new method bodies are perf regressions (possibly for jit/AOT, and almost certainly for the interpreter)

/cc @kg

**Comment by stephentoub** [comment]
> Are there new/existing tests for the throwing behavior?

Yes:
https://github.com/dotnet/runtime/pull/104558/files#diff-b7ba854effe0b40e624638bc9ea70e50950841a776d09a3a1f36df3922f9efdfR397
https://github.com/dotnet/runtime/pull/104558/files#diff-b7ba854effe0b40e624638bc9ea70e50950841a776d09a3a1f36df3922f9efdfR824

> If mono is green, one of two things is wrong

The mono legs are green (or at least the failures in the wasm legs are unrelated).

Are you saying something is definitely wrong, or are you saying _if_ something were wrong and going unnoticed it would be those?

From what I could tell, mono's intrinsic implementations here already only support a subset of what the APIs do and already fall back to the managed implementations for other cases, and the throwing behavior is handled in the managed implementation. Did I misunderstand how things fit together?

**Comment by lambdageek** [comment]
> The mono legs are green (or at least the failures in the wasm legs are unrelated).

> Are you saying something is definitely wrong, or are you saying if something were wrong and going unnoticed it would be those?

I think we should keep an eye on the perf runs - I wonder if generic sharing will see a perf regression. 
We're giving the inlining heuristic more work to do. The managed body of `Interlocked.CompareExchange<T>` is bigger now, so if the initial codegen doesn't remove the dead branches effectively it might inhibit inlining the callers of `CompareExchange<T>` into _their_ callers.

Also startup perf on the interpreter might be affected since it doesn't do many optimizations in Tier0.

> From what I could tell, mono's intrinsic implementations here already only support a subset of what the APIs do and already fall back to the managed implementations for other cases, and the throwing behavior is handled in the managed implementation. Did I misunderstand how things fit together?

No, I think it's fine. 

In JIT/AOT we handle two cases: `CompareExchange(ref X, X, X)` where `X` is int, long, float, double, intptr and object; and `CompareExchange(ref object, ref object, ref object, ref object)` (The latter is an implementation detail of `CompareExchange(ref object, object, object)`).  It will fall back to the managed implementation for: unsigned primitives, enums, other valuetypes, and generic sharing / gsharedvt cases (for generic sharing, the codegen will see a generic parameter with a special flag which won't match any of the cases that the intrinsic wants to handle).  I think with the exception of gsharedvt, all the other cases should fold the `typeof(T).IsValueType` and `sizeof(T) == N` tests during the initial codegen.  (but if it doesn't then we'd see perf regressions).   (there is a third kind of generic sharing that sometimes shows up: partial generic sharing - this tries to share code for `int` and "enum with underlying type == int" (and long, short, byte) - I think those cases should work like gshared for reference types - all the false branches should be dropped during initial codegen)

In the interpreter we only seem to handle `int` and `long` and run the managed versions for everything else (but the interpreter also doesn't do generic sharing, so we only will see concrete instantiations).  In general startup perf suffers if there's more method calls on hot paths since tier0 doesn't do as much optimization and managed->managed method calls are expensive in tier0.

tl;dr:

1. we should measure startup perf on the interpreter and see if there's any ill effects
2. we should look at the generated code for gshared and enum inputs in the JIT/AOT and verify that all the `typeof(T).IsValueType` and `sizeof(T) == N` branches are being removed in the initial codegen so that downstream optimizations aren't affected

**Comment by lambdageek** [comment]
oh, in case it wasn't clear: I don't think we should hold up this PR due to https://github.com/dotnet/runtime/pull/104558#issuecomment-2236776021 - just that we have some follow-up work

**Comment by stephentoub** [comment]
Thanks, @lambdageek.

In that case, I think the remaining blocker is just figuring out why I'm getting some invalid program exceptions with native aot :)

**Comment by MichalStrehovsky** [comment]
> In that case, I think the remaining blocker is just figuring out why I'm getting some invalid program exceptions with native aot :)

Pushed out a fix for that. There was also a problem in crossgen2 but there it was just a deoptimization since we skip compiling methods with invalid IL and leave it to JIT.

**Comment by stephentoub** [comment]
> > In that case, I think the remaining blocker is just figuring out why I'm getting some invalid program exceptions with native aot :)
> 
> Pushed out a fix for that. There was also a problem in crossgen2 but there it was just a deoptimization since we skip compiling methods with invalid IL and leave it to JIT.

Thank you, @MichalStrehovsky! It would have taken me a while to discover the existence of TryGetIntrinsicMethodIL vs TryGetPerInstantiationIntrinsicMethodIL.

**Review by jkotas — APPROVED** [review]
LGTM!

**Comment by stephentoub** [comment]
Thanks, all.

**Pull request merged**
